### PR TITLE
CallCacheActor skeleton strawman

### PR DESCRIPTION
**Pull request by cjllanwarne**
Something of a strawman commit. Very roughly maps out how I think we said the CallCacheActor would fit into the existing actor model?

### CODE REVIEW

**Comment by geoffjentry** [comment]
I just skimmed this, but it looks like it's more or less what I was talking about when we chatted earlier. 

OTOH the reason I further delayed writing up the ticket was I had some thoughts when I was writing them that I thought were worth talking about, but no one is here at the moment :)

**Comment by geoffjentry** [comment]
I'm guessing that the "people aren't around" thing isn't going to change at 5pm on a Friday, so ....

I'm thinking now that instead, perhaps the thing to do is to also have the actor know if read is on and if write is on. When the actor spins up, it ...

- If read is on, just starts determining the cache hit status w/o being asked
- If write is on, starts hashing everything
- If both are on, starts hashing everything but is checking for cache hit until a miss occurs, at which point it's just generating hashes

The actor could then receive two messages, one is "are you a cache hit" and the other is "please now persist to the store" (if you asked for cache hit status before it was done, perhaps it could send back a "come back later" message - that way you can avoid returning the Futures, although perhaps people like that)

We know for a fact that if read is on that we'll be checking the cache, so might as well start that ASAP. If write is on, we're potentially wasting energy - e.g. if a job fails, but presumably (hopefully!) job successes are more common than job failures and we can get a jumpstart on the process.

**Comment by francares** [comment]
@geoffjentry and @cjllanwarne Could you provide a link to CallCache design documentation?... Thanks.

**Comment by cjllanwarne** [comment]
@geoffjentry I like that - don't layer on more futures, just start working and either say "ready and here's the answer" or "not ready yet, ask me later". It's also much more RESTy (I refuse to say `RESTful`) like that, which is probably a good sign

**Comment by cjllanwarne** [comment]
Just FYI, the `BackendJobExecutionActor` is still a work in progress ATM. Changes incoming

File: backend/src/main/scala/cromwell/backend/BackendJobExecutionActor.scala [hunk 1]
```diff
@@ -33,8 +35,17 @@ object BackendJobExecutionActor {
   */
 trait BackendJobExecutionActor extends BackendJobLifecycleActor with ActorLogging {
 
+  val callCacheActor = context.actorOf(CallCachingActor.props(jobDescriptor, _ => "Implement me!", _ => "Implement me!"))
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/BackendJobExecutionActor.scala]: Technically those are implemented, they're just terrible hashing functions.

File: backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    case Event(CheckCache, _) =>
+      sender ! CallCacheCheckInProgress
+      stay()
+    case Event(CacheHitInternal(cacheResultId), _) if readFromCache =>
+      readEntryFromDatabase(cacheResultId)
+      goto(FetchingJobResult)
+    case Event(CacheHitInternal(_), _) if !readFromCache =>
+      goto(UnreadCacheHit)
+    case Event(CacheMissInternal(hashes), _) if writeToCache =>
+      goto(AwaitingJobResultAndAllHashes) using WithHashes(hashes)
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala]: Not clear why this sets the state data to `WithHashes(hashes)`?  It doesn't look like this is considered in `AwaitingJobResultAndAllHashes`.

File: backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    case Event(CacheMissInternal(_), _) if !writeToCache =>
+      goto(UnupdatedCacheMiss)
+  }
+
+  when(AwaitingJobResultAndAllHashes) {
+    case Event(CheckCache, _) =>
+      sender ! CallCacheMiss
+      stay()
+    case Event(WriteCache(returnCode, jobOutputs), WithHashes(hashes)) =>
+      writeEntryToDatabase(returnCode, jobOutputs)
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala]: maybe include the hashes in what gets written?  😄

File: backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+      stay()
+    case Event(CheckWriteComplete, _) =>
+      sender ! CallCacheWriteInProgress
+      stay()
+    case Event(WriteCache(_, _), _) =>
+      // For the sake of idempotency, respond to this message too:
+      sender ! CallCacheWriteInProgress
+      stay()
+    case Event(CallCacheWriteSuccessInternal, oldData @ WithJobResult(_, _)) =>
+      goto(NewCacheEntryWritten) using oldData
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala]: what's going on with `oldData`?

File: backend/src/main/scala/cromwell/backend/BackendJobExecutionActor.scala [hunk 5]
```diff
@@ -33,8 +35,19 @@ object BackendJobExecutionActor {
   */
 trait BackendJobExecutionActor extends BackendJobLifecycleActor with ActorLogging {
 
+  val callCacheActor = context.actorOf(SyncCallCachingActor.props(jobDescriptor, _ => "Implement me!", _ => "Implement me!", readFromCache = false, writeToCache = false))
+
+  private var executeClient: Option[ActorRef] = None
```
Comment (francares) [inline backend/src/main/scala/cromwell/backend/BackendJobExecutionActor.scala]: Why this var is needed?

File: backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    case Event(CheckCache, _) =>
+      sender ! CallCacheMiss
+      stay()
+    case Event(WriteCache(_, _), _) =>
+      sender ! CallCacheWriteSuccess
+      stay()
+  }
+
+  private def calculateHash(): Unit = {
+    // TODO: Do the work and then:
```
Comment (francares) [inline backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala]: The way of calculating hash will be the same for all backends?

File: backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    // self ! CacheHitInternal(cacheResultId)
+    // or collate all the hashes and then:
+    self ! CacheMissInternal(Map.empty)
+  }
+
+  private def writeEntryToDatabase(returnCode: Option[Int], jobOutputs: JobOutputs): Unit = {
+    // TODO: Do the work and then:
```
Comment (francares) [inline backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala]: Same question here but regarding database.

File: backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala (hunk 6 above)
Comment (geoffjentry) [inline backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala]: If a backend chooses to use this call caching & except for runtime attributes and files, yes.

File: backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala (hunk 7 above)
Comment (geoffjentry) [inline backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala]: Not if a backend chooses to use this call caching. The intention was that it was as on rails as possible. A backend can choose to use this call caching out of the box (as much as possible, some things are inherently backend specific) or roll their own.

**Comment by cjllanwarne** [comment]
This strawman was brave... but it wasn't good enough. Since this no longer represents our current thinking I'm going to take this strawman down.

**Comment by mcovarr** [comment]
RIP strawman 😢

**Pull request closed without merging**